### ISSUE TITLE
[$1000] Web - Thread header with “From undefined (archived)“ displayed for a while after leaving room

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#22435]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to the “New room” 
2. Create a new public room
3. Send a message
4. Click on “Reply in thread” 
5. Pin the thread message for quick access 
6. Open room details and leave the room
7. Try to open the thread message again quickly

## Expected Result:
Thread header with “From undefined (archived)”should not be displayed

## Actual Result:
Thread header with “From undefined (archived) displayed for a while

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.37-5
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/6315a383-c564-45ac-8dec-76ce3afe3e14

https://github.com/Expensify/App/assets/93399543/42b03dcb-82b6-4d97-b8b1-6c40829196f5

**Expensify/Expensify Issue URL:**
**Issue reported by:** @ayazhussain79
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1688676493813539

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01455d688d5910102c</li>
        <li>Upwork Job ID: 1678418170472640512</li>
        <li>Last Price Increase: 2023-07-31</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @francoisl</details>

**Comment by manjesh-yadav** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Thread header with “From undefined (archived)“ displayed for a while after leaving room
### What is the root cause of that problem?

In javascript when we concatenate an `undefined` variable with a string it converts to an `'undefined' string` this is done on function `getReportName`
https://github.com/Expensify/App/blob/957c2e0058dafd400362c891cc7b17b8be836a78/src/libs/ReportUtils.js#L1118-L1120

that is causing a problem and we failed to check if the report name is empty or not in function `getParentNavigationSubtitle` line 1198 and we return an undefined string

https://github.com/Expensify/App/blob/957c2e0058dafd400362c891cc7b17b8be836a78/src/libs/ReportUtils.js#L1194-L1205



### What changes do you think we should make in order to solve the problem?

Before concatenating we should check if the variable is not `undefined`

```diff
`
-if (isArchivedRoom(report)) {
+if (isArchivedRoom(report)  && !_.isUndefined(formattedName)) {
        formattedName += ` (${Localize.translateLocal('common.archived')})`;
}
```
When we leave the room and quickly switch its child thread, that time parent report is null that is also causes a problem and throws an error on the code

https://github.com/Expensify/App/blob/957c2e0058dafd400362c891cc7b17b8be836a78/src/libs/ReportUtils.js#L1158-L1161

Hare, we should have to check if `report` is not null or is an object before calling `getPolicyName(report, true)`

```js
return {
        rootReportName: _.isObject(report) ? getReportName(report) : '',
        workspaceName: _.isObject(report) ? getPolicyName(report, true) : '',
    };
```

**Comment by abdulrahuman5196** [comment]
Hi, will complete the review in the morning.

**Comment by laurenreidexpensify** [comment]
@abdulrahuman5196 have you had a chance to review? thanks

**Comment by abdulrahuman5196** [comment]
From @hungvu193 's proposal as well as @manjesh-yadav . I am not aligned on the root cause nor the solution.
Both the solution is concerned about the case of making the `undefined` value work. But it doesn't provide information on why it comes for sometime and disappears(I assume it has to do with optimistic data or openReport API call).

Curious question @laurenreidexpensify , after leaving a room we remove the room information from the user, then what is the point of showing the thread alone?

**Comment by manjesh-yadav** [comment]
Yes @abdulrahuman5196, It is because of optimistic data and `LeaveRoom` API call

When we call `LeaveRoom` we set optimistic data which make parent report of that thread like `{stateNum: 1, statusNum: 1}` and after API success this object data will be deleted

https://github.com/Expensify/App/blob/957c2e0058dafd400362c891cc7b17b8be836a78/src/libs/actions/Report.js#L1701

```js
stateNum: CONST.REPORT.STATE_NUM.SUBMITTED,
statusNum: CONST.REPORT.STATUS.CLOSED,
```

**Removing this optimistic data will also solve our problem because removing this we do not have a parent report at the moment.**



> Curious question @laurenreidexpensify , after leaving a room we remove the room information from the user, then what is the point of showing the thread alone?

We have separate option for leave thread

**Comment by abdulrahuman5196** [comment]
Will check this in a while

**Comment by abdulrahuman5196** [comment]
Checking now

**Comment by abdulrahuman5196** [comment]
@manjesh-yadav Thanks for the information here - https://github.com/Expensify/App/issues/22435#issuecomment-1642278597
Kindly add the information to your proposal. I dont agree on removing optimistic data.

**Comment by manjesh-yadav** [comment]
> @manjesh-yadav Thanks for the information here - [#22435 (comment)](https://github.com/Expensify/App/issues/22435#issuecomment-1642278597) Kindly add the information to your proposal. I dont agree on removing optimistic data.

@abdulrahuman5196 proposal updated with more information [Proposal](https://github.com/Expensify/App/issues/22435#issuecomment-1625975684)

**Comment by laurenreidexpensify** [comment]
@trjExpensify @grgia can you help me unravel if this is expected behaviour?

**Comment by trjExpensify** [comment]
Not sure exactly what parts you want input on, but a workspace name in a chat title shouldn't show as `undefined` anywhere really.

We had a similar problem before with the workspace name on the workspace chats, #admins/#announce rooms showing as `undefined` when archived, but we fixed those back then. CC: @mountiny

**Comment by abdulrahuman5196** [comment]
@trjExpensify 

The question was these 2

1) Curious question, after leaving a room we remove the room information totally from the user, then what is the point of showing the thread alone?

2) If 1 is the expected to show, We will need to actually remove the 'From ..' text in header which is displayed for a while after leaving room, similar to how its removed after a while. But we won't be able to show the room name(Because that data is removed). Is this fine?

**Comment by laurenreidexpensify** [comment]
@trjExpensify is OOO today will ask him to look again on Monday

**Comment by mountiny** [comment]
@trjExpensify When we leave room I assume we still stay in the threads.

I think the solution would be loop over all threads we have locally where this room is parentReportID and remove this optimistically from the onyx data

**Comment by laurenreidexpensify** [comment]
@mountiny do you know if this is definitely the design flow? (I'm not sure myself, I just wanted to confirm if this is a guess or a definite thing) 

> When we leave room I assume we still stay in the threads.

**Comment by trjExpensify** [comment]
Hm, this is probably quite involved across a number of different consideration for user-created rooms isn't it? Some of which are probably pending for future iterations of functionality like search/discovery and keyword search of rooms/threads you were once a member of and can re-join/find content in. At which point, the chat header of a thread should always show the parent room name and workspace name regardless of your current room membership status. 

I agree there's a question here first on automatically unsubscribing from the threads of a parent room you leave that could potentially avoid it from being a consideration for the time being, but we're going to run into it again soon enough even if that's intended I think. For example.. 

1. Join a #social room on the Duraflame workspace, subscribe to a thread. (Chat header: "From #social in Duraflame")
2. Leave the room and unsubscribe from the thread
3. search for keywords contained in a message within the thread
4. access the thread to view the content returned in my search. (Chat header: "From undefined in undefined"?)

Anyhow, let's ask @MitchExpensify & @jasperhuangg on expected behaviour when leaving a room today as I'm not sure where that's at right now. Specifically, what should happen with the child threads of the rooms you leave? 

Option A: 
- If you leave the room, the room disappears from your LHN
- You're automatically unsubscribed from the room's threads, so those disappear as well
- The room isn't discoverable again via search, neither are the room's threads (only via a QR code or join link etc)

Option B:
- If you leave the room, it disappears from your LHN
- You remain subscribed to the threads of the room you left (thus they remain in your LHN and searchable etc)

**Comment by MitchExpensify** [comment]
This isn't covered in the User Created Rooms Release project so we'd need to see what direction we'd like to go in #product or #expensiy-open-source. 

I think Option B makes sense at this point. Each thread is treated as its own chat/report so I'm not sure I'd expect those to become unsearchable when I leave a room.

**Comment by jasperhuangg** [comment]
Option B is the current design, since we treat threads as their own reports, leaving the parent report means you're still a member of the thread reports. I think that's how we should leave things for now.

**Comment by trjExpensify** [comment]
I'm not sure I totally agree with Option B, but if that's the intention right now, we need to fix this `undefined` bug in the chat header. As I mentioned [earlier in the thread](https://github.com/Expensify/App/issues/22435#issuecomment-1667904978), we'll come across this again anyhow. 👍

**Comment by abdulrahuman5196** [comment]
@trjExpensify 

Just to confirm again since there are couple discussion and quoting my question from here https://github.com/Expensify/App/issues/22435#issuecomment-1660489336 again.

>We will need to actually remove the 'From ..' text in header which is displayed for a while after leaving room, similar to how its removed after a while. But we won't be able to show the room name(Because that data is removed). Is this fine?

So for the threads from a room which user has already left, we won't show 'From ..' header. This is correct?

**Comment by trjExpensify** [comment]
>So for the threads from a room which user has already left, we won't show 'From ..' header. This is correct?

I think it's weird for a staple portion of the chat header to be there sometimes and other times not. I would suggest the From line remains, but the parent chat portion is no longer hyperlinked if you don't have access to the parent chat anymore.

**Comment by jasperhuangg** [comment]
@trjExpensify I honestly agree that we should probably change how that works. If you're not in a room anymore it doesn't make sense for you to be in any of its threads.

@chiragsalian what do you think? Was this talked about at all during the threads design?

**Comment by laurenreidexpensify** [comment]
It sounds like this is the expected behaviour we agree on - @trjExpensify @MitchExpensify @jasperhuangg @mountiny ? 

> the From line remains, but the parent chat portion is no longer hyperlinked if you don't have access to the parent chat anymore.

Can you 👍 if you agree

**Comment by laurenreidexpensify** [comment]
@abdulrahuman5196 confirmation above ^^ thanks

**Comment by laurenreidexpensify** [comment]
@abdulrahuman5196 please note confirmation above

**Comment by abdulrahuman5196** [comment]
@laurenreidexpensify @trjExpensify  AFAIK we don't have the original room details to show in the header itself, Like the room name of the thread. All the room details are removed once the room is left. 

Correct me if wrong @jasperhuangg @hungvu193

**Comment by abdulrahuman5196** [comment]
Then we have to look ways to retain the room name.

**Comment by trjExpensify** [comment]
^^ Yeah, exactly.

**Comment by jasperhuangg** [comment]
@abdulrahuman5196 Ah yes that's a correct assumption. It seems we need some sort of back-end change to achieve this, however, i.e. we'll need to start sending the parent report's information with child reports, so that when the parent report gets deleted we still have the data we need?

**Comment by laurenreidexpensify** [comment]
@trjExpensify @MitchExpensify @jasperhuangg @mountiny would you agree that this needs to be an internal issue rather than external at this point?

**Comment by trjExpensify** [comment]
Yep agreed, `internal` for the backend changes. :)

**Comment by abdulrahuman5196** [comment]
Yes agreed we should start from internal. It would be better to take this internal and decide on how we should retain the data and then if required we can open some parts of changes to external.

**Comment by laurenreidexpensify** [comment]
@francoisl after a bit of back and forth this was determined to be a back end issue before we can open up to external

**Comment by abdulrahuman5196** [comment]
@francoisl Bump.

**Comment by francoisl** [comment]
Started to look into this, but heads up I will need a few more days - too many competing priorities and this doesn't look like the most urgent issue right now.

**Comment by laurenreidexpensify** [comment]
waiting on @francoisl to have bandwidth to dig into

**Comment by laurenreidexpensify** [comment]
@francoisl will you be able to look into this one this week?

**Comment by laurenreidexpensify** [comment]
@francoisl bump - should we make this a weekly?

**Comment by francoisl** [comment]
Yes, I haven't had time to look this week.

**Comment by laurenreidexpensify** [comment]
@francoisl any capacity yet?

**Comment by laurenreidexpensify** [comment]
@francoisl pls update thanks

**Comment by francoisl** [comment]
Correct, I don't have the bandwidth to work on this at the moment and it doesn't sound like a very bad bug.

**Comment by francoisl** [comment]
Still haven't had a chance to look into this 😓

**Comment by francoisl** [comment]
Still haven't had a chance to look into this

**Comment by francoisl** [comment]
Still haven't prioritized this

**Comment by laurenreidexpensify** [comment]
Let's close this and reopen if it becomes a more widespread issue affecting more users. It feels edge case for now and low priority.
